docs: add Ts.ED example usage (#2587)
mikro-orm/mikro-orm: docs/docs/usage-with-tsed.md

---
title: Using MikroORM with Ts.ED framework
sidebar_label: Usage with Ts.ED
---

## Installation

> Easiest way to integrate MikroORM to Ts.ED is via [`@tsed/mikro-orm` module](https://tsed.io/tutorials/mikroorm.html).

Reviewer: there should be some clear note that there is possibly more up to date version of this recipe in the tsed docs

Author: Isn't necessary. The API have changed on our side. The latest changes is internal and your documentation point to the latest release. It shouldn't be a problem for you.

Author: @derevnjuk can you confirm this point ?

Reviewer 2: @Romakita sure, you are right. However, I guess @B4nan  meant that we should point out somehow that source of truth is our docs.

Author: Ok. Do you have a formulation or an example of a sentence that would suit you in this case?

As I don't see anything equivalent on other pages such as Nest.js and Admin.js, I wouldn't want to put anything :)

Suggestion:

```
::: tip Note
The documentation presented on this page is taken from the Ts.ED site. The latest updates to this example can be found directly on the [`@tsed/mikro-orm` module](https://tsed.io/tutorials/mikroorm.html) page.
:::
```

Reviewer: The nest.js guide has such link in the opposite direction, as the `@mikro-orm/nestjs` is our package. I'd imagine something like that. 

<img width="1061" alt="image" src="https://user-images.githubusercontent.com/615580/179505893-ade5ef67-9ea1-482a-b492-f25eae757228.png">

https://docs.nestjs.com/recipes/mikroorm

Author: done. Sorry for the very late answer @B4nan.

$ yarn add @mikro-orm/core @tsed/mikro-orm @mikro-orm/mysql       # for mysql/mariadb
$ yarn add @mikro-orm/core @tsed/mikro-orm @mikro-orm/mariadb     # for mysql/mariadb
$ yarn add @mikro-orm/core @tsed/mikro-orm @mikro-orm/postgresql  # for postgresql
$ yarn add @mikro-orm/core @tsed/mikro-orm @mikro-orm/sqlite      # for sqlite
```

or

```bash
$ npm i -s @mikro-orm/core @tsed/mikro-orm @mikro-orm/mongodb     # for mongo
$ npm i -s @mikro-orm/core @tsed/mikro-orm @mikro-orm/mysql       # for mysql/mariadb
$ npm i -s @mikro-orm/core @tsed/mikro-orm @mikro-orm/mariadb     # for mysql/mariadb
$ npm i -s @mikro-orm/core @tsed/mikro-orm @mikro-orm/postgresql  # for postgresql
$ npm i -s @mikro-orm/core @tsed/mikro-orm @mikro-orm/sqlite      # for sqlite
```

Once the installation process is completed, we can import the `MikroOrmModule` into the `Server` configuration:

```typescript
import { Configuration } from '@tsed/di';
import { MikroOrmModule } from '@tsed/mikro-orm';

@Module({
  imports: [MikroOrmModule],
  mikroOrm: [
    {
      contextName: 'default',
      type: 'postgres',
      ...,

      entities: [

>         './src/entity/*{.ts,.js}'

Reviewer: trailing comma
```suggestion
        './src/entity/*{.ts,.js}',
```

})
export class Server {}
```

The `mikroOrm` options accepts the same configuration object as `init()` from the MikroORM package. Check [this page](configuration.md) for the complete configuration documentation.

## Obtain ORM instance

`@Orm` decorator lets you retrieve an instance of MikroOrm.

```ts
import { Injectable } from '@tsed/di';
import { Orm } from '@tsed/mirkro-orm';

@Injectable()
export class MyService {
  @Orm()
  private readonly orm!: MikroORM;

  async create(user: User): Promise<User> {

    // do something
    // ...
    // Then save
    await this.orm.em.persistAndFlush(user);
    console.log('Saved a new user with id: ' + user.id);

    return user;
  }

  async find(): Promise<User[]> {
    const users = await this.orm.em.find(User, {});
    console.log('Loaded users: ', users);

    return users;
  }
}
```

It's also possible to inject an ORM instance by its name:

```ts
import { Injectable } from '@tsed/di';
import { Orm } from '@tsed/mirkro-orm';

@Injectable()
export class MyService {
  @Orm('mongo')
  private readonly orm!: MikroORM;
}
```

## Use Entity with Controller

To begin, we need to define an Entity MikroORM to define a model:

```typescript

> import { Entity, Property, PrimaryKey, Property } from "@mikro-orm/core";

Reviewer: ```suggestion
import { Entity, Property, PrimaryKey, Property } from '@mikro-orm/core';
```

> import { Required } from '@tsed/schema'

Reviewer: ```suggestion
import { Required } from '@tsed/schema';
```

Extra Ts.ED decorators are available to validate model and/or generate Swagger document. See more details on the dedicated 
documentation page [here](https://tsed.io/docs/model.html).

We can use this model with a Controller like that:

```typescript
import { Controller, Post, BodyParams, Inject, Post, Get } from '@tsed/common';

@Controller('/users')
export class UsersCtrl {
  @Inject()
  private usersService: UsersService;

  @Post('/')
  create(@BodyParams() user: User): Promise<User> {
    return this.usersService.create(user);
  }

  @Get('/')
  getList(): Promise<User[]> {
    return this.usersService.find();
  }
}
```

## Obtain EntityManager

`@EntityManager` and `@Em` decorators lets you retrieve an instance of EntityManager.

```typescript
import { Injectable } from '@tsed/di';
import { Em } from '@tsed/mikro-orm';
import { EntityManager } from '@mikro-orm/mysql'; // Import EntityManager from your driver package or `@mikro-orm/knex`

@Injectable()
export class UsersService {
  @Em()
  private readonly em!: EntityManager;

  async create(user: User): Promise<User> {
    await this.em.persistAndFlush(user);

>     console.log("Saved a new user with id: " + user.id);

Reviewer: ```suggestion
    console.log('Saved a new user with id: ' + user.id);
```
